File: web/src/lib/attention.test.ts
import { describe, expect, it } from 'vitest'
import type { AppListEntry } from '../types/appDetail'
import type { NodeResource } from '../types/nodeDetail'
import { buildAttentionItems } from './attention'

const app = (name: string, variant: 'success' | 'destructive' | 'muted') =>
  ({ name, status: { label: 'x', variant } }) as unknown as AppListEntry
const node = (id: string, status: NodeResource['status']) =>
  ({ id, name: id, status }) as unknown as NodeResource

describe('buildAttentionItems', () => {
  it('returns nothing when everything is healthy', () => {
    expect(
      buildAttentionItems({
        apps: [app('a', 'success'), app('b', 'muted')],
        nodes: [node('n1', 'online'), node('n2', 'cordoned')],
        certs: [
          {
            domain: 'x.io',
            status: 'healthy',
            not_before: '',
            not_after: '',
          },
        ],
        doctor: {
          ok: true,
          checks: [{ code: 'c', name: 'c', status: 'ok', message: '' }],
        },
      }),
    ).toEqual([])
  })

  it('flags failing apps, offline nodes, bad certs and doctor problems, critical first', () => {
    const items = buildAttentionItems({
      apps: [app('web', 'destructive')],
      nodes: [node('n1', 'offline')],
      certs: [
        {
          domain: 'a.io',
          status: 'expiring_soon',
          not_before: '',
          not_after: '2026-10-01',
        },
        {
          domain: 'b.io',
          status: 'expired',
          not_before: '',
          not_after: '2026-09-01',
        },
      ],
      doctor: {
        ok: false,
        checks: [
          { code: 'disk', name: 'Disk', status: 'warn', message: 'low' },
          { code: 'docker', name: 'Docker', status: 'fail', message: 'down' },
          { code: 'net', name: 'Net', status: 'unknown', message: '' },
        ],
      },
    })
    expect(items.map((i) => i.id)).toEqual([
      'app:web',
      'node:n1',
      'cert:b.io',
      'doctor:docker',
      'cert:a.io',
      'doctor:disk',
    ])
    expect(items.filter((i) => i.severity === 'critical')).toHaveLength(4)
  })

  it('flags disk pressure', () => {
    const items = buildAttentionItems({
      disk: {
        level: 'critical',
        freePercent: 3,
        freeBytes: 3e9,
        reclaimableBytes: 1e9,
      },
    })
    expect(items).toHaveLength(1)
    expect(items[0]).toMatchObject({ id: 'disk', severity: 'critical' })
  })

  it.each([
    {
      name: 'with a rollback target',
      last_good_image: 'web:1',
      error: 'build failed',
      wantDetail: 'build failed',
      wantLastGood: 'web:1',
    },
    {
      name: 'without a rollback target or error',
      last_good_image: undefined,
      error: undefined,
      wantDetail: 'No error recorded',
      wantLastGood: undefined,
    },
  ])('flags a failed deploy $name', (c) => {
    const items = buildAttentionItems({
      failedDeploys: [
        {
          id: 'dep_2',
          service_name: 'web',
          image: 'web:2',
          status: 'failed',
          started_at: new Date().toISOString(),
          error: c.error,
          last_good_image: c.last_good_image,
        },
      ],
    })
    expect(items).toHaveLength(1)
    expect(items[0]).toMatchObject({
      id: 'deploy:web',
      severity: 'critical',
      target: {
        kind: 'deploy',
        app: 'web',
        image: 'web:2',
        lastGoodImage: c.wantLastGood,
      },
    })
    expect(items[0].detail).toContain(c.wantDetail)
  })
})

File: cmd/levelrail-cli/attention_test.go
package main

import (
	"testing"
	"time"

	"github.com/GLINCKER/levelrail/internal/apiclient"
)

func appEntry(name, variant string) apiclient.AppStatusEntry {
	var e apiclient.AppStatusEntry
	e.Name = name
	e.Status.Label = "Attention needed"
	e.Status.Variant = variant
	return e
}

func TestBuildAttentionItems(t *testing.T) {
	t.Parallel()
	notAfter := time.Date(2026, 10, 1, 0, 0, 0, 0, time.UTC)
	tests := []struct {
		name   string
		apps   []apiclient.AppStatusEntry
		nodes  []nodeResource
		certs  []apiclient.CertificateResource
		doc    systemDoctorResource
		failed []apiclient.FailedDeployResource
		status apiclient.SystemStatusResource
		want   []string
	}{
		{name: "healthy", apps: []apiclient.AppStatusEntry{appEntry("a", "success")}, want: []string{}},
		{
			name:  "critical sorts before warning",
			apps:  []apiclient.AppStatusEntry{appEntry("web", "destructive")},
			nodes: []nodeResource{{Name: "n1", Status: "offline"}, {Name: "n2", Status: "cordoned"}},
			certs: []apiclient.CertificateResource{
				{Domain: "a.io", Status: "expiring_soon", NotAfter: notAfter},
				{Domain: "b.io", Status: "expired", NotAfter: notAfter},
			},
			doc: systemDoctorResource{Checks: []doctorCheckResource{
				{Name: "Disk", Status: "warn"},
				{Name: "Docker", Status: "fail"},
				{Name: "Net", Status: "unknown"},
			}},
			want: []string{"critical:web", "critical:n1", "critical:b.io", "critical:Docker", "warning:a.io", "warning:Disk"},
		},
		{
			name:   "failed deploy and critical disk sort first",
			failed: []apiclient.FailedDeployResource{{DeployAttemptResource: apiclient.DeployAttemptResource{ServiceName: "api", Error: "build failed"}}},
			status: apiclient.SystemStatusResource{DataDirTotalBytes: 100, DataDirFreeBytes: 4},
			want:   []string{"critical:data dir", "critical:api"},
		},
		{
			name:   "warning disk at 8 percent free",
			status: apiclient.SystemStatusResource{DataDirTotalBytes: 100, DataDirFreeBytes: 8},
			want:   []string{"warning:data dir"},
		},
		{
			name:   "disk at 10 percent free is fine",
			status: apiclient.SystemStatusResource{DataDirTotalBytes: 100, DataDirFreeBytes: 10},
			want:   []string{},
		},
		{
			name:   "unknown disk size is ignored",
			status: apiclient.SystemStatusResource{},
			want:   []string{},
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			got := buildAttentionItems(attentionInput{Apps: tc.apps, Nodes: tc.nodes, Certs: tc.certs, Doctor: tc.doc, Failed: tc.failed, Status: tc.status})
			if len(got) != len(tc.want) {
				t.Fatalf("got %d items, want %d: %+v", len(got), len(tc.want), got)
			}
			for i, it := range got {
				if key := it.Severity + ":" + it.Subject; key != tc.want[i] {
					t.Errorf("item %d = %s, want %s", i, key, tc.want[i])
				}
			}
		})
	}
}
